### PR TITLE
WIP: Add migration feature

### PR DESCRIPTION
**Pull request by KABBOUCHI**
Create migration when create/update/delete table

TODO:
- Create migration for BREAD actions
- ~~Insert current migration history~~
- Enable database tool only for local env (optional)
- Tests
- Refactor and extract common code

 Any thought before I continue?

### CODE REVIEW

File: src/Http/Controllers/VoyagerDatabaseController.php [hunk 1]
```diff
@@ -59,6 +56,40 @@ public function create()
         return Voyager::view('voyager::tools.database.edit-add', compact('db'));
     }
 
+    protected function prepareDbManager($action, $table = '')
```
Comment (fletch3555) [inline src/Http/Controllers/VoyagerDatabaseController.php]: Any particular reason this method was moved?

Comment (KABBOUCHI) [reply src/Http/Controllers/VoyagerDatabaseController.php]: PHPStorm auto format on save

**Comment by fletch3555** [comment]
Liking the idea so far. We just need to make sure it fully supports multiple database systems (I believe mysql/Mariadb and postures are the only 2 Voyager supports, so that would be a good start), including indices, foreign keys, etc.  For Voyager specifically, we also need to implement this for BREAD create/modify/delete (data_type and data_rows tables)

This would make for a very good package once complete.  Then Voyager would only need to implement the appropriate events, and the package could have a configuration tying package listeners to Voyager (or whatever project) events.  The hardest part is going to be making sure you fully support multiple DBMS

**Comment by addingama** [comment]
@KABBOUCHI it looks like there is a conflict that you need to fix


I hope this can be merged soon

**Comment by KABBOUCHI** [comment]
@dashracer Sorry, I was busy this month, I'm planning to rebase and retargeting to `1.x-dev` branch and finishing the rest in the next few weekends.

**Comment by fletch3555** [comment]
Shouldn't need to rebase or open a new PR.  Just pull 1.x-dev into your branch and fix the conflicts.  You'll have conflicts either way you do it.  This way just saves us from duplicate PRs

**Comment by alrik11es** [comment]
Are those problems just styling problems?
This is an interesting feature.

**Comment by yasseralsamman** [comment]
when creating a migration for BREAD update, the getMigrationUpCommand function is depending on the data type id, but this id might change if you are migrating to a new database and will raise the error:
`Call to a member function updateDataType() on null`

I suggest depending on the name instead
in `CreateBreadUpdateMigration.php` line 52:
```
        $content = str_replace(
            '//MIGRATION_UP',
            $this->getMigrationUpCommand($dataType->name, $event->request_data),
            $content
        );
```
then in line 97:
`        $dataType = Voyager::model("DataType")->where("name","' . $dataTypeName . '")->first();
`

this should also be done in `CreateBreadDeleteMigration.php`

**Comment by NicholasDShaw** [comment]
Any chance this is still being worked on? Love voyager after playing with it a bit, but feel that this feature is really a big deal in choosing it to build an application.

**Comment by SidharthRaveendran** [comment]
Much required feature, is this still in works?

**Comment by KABBOUCHI** [comment]
@SidharthRaveendran I think I'm gonna split this PR into 2, table migration and BREAD migration, the table migration should be easy and ready to be merged ASAP.

**Comment by impsart** [comment]
Following. Looking forward to this.

**Comment by gerardreches** [comment]
Was this finally implemented?

**Comment by emptynick** [comment]
Closing due to inactivity.

Please feel free to reopen or grab this fork and continue developing.
But note that this will likely be unnecessary for Voyager 2.

**Pull request closed without merging**